File: sdk/flutter/lib/src/customer_restore_sync.dart
import 'dart:async';

import 'commerce.dart';
import 'customer_entitlement_runtime.dart';
import 'customer_entitlements.dart';
import 'transaction_observation.dart';

/// Mosaic's authoritative answer to a restore. It is separate from what the
/// native provider did, because a successful native restore whose facts have
/// not been validated is not restored access, and reporting it as one is how a
/// restore flow starts lying.
enum MosaicCustomerRestoreOutcome {
  restored('restored'),
  noAdditionalPurchases('no_additional_purchases'),
  validationPending('validation_pending'),
  identityUnresolved('identity_unresolved'),
  productUnresolved('product_unresolved'),
  providerUnavailable('provider_unavailable'),
  failed('failed');

  const MosaicCustomerRestoreOutcome(this.wireValue);

  final String wireValue;
}

/// What the native provider restore itself did, reported separately and never
/// merged into the authoritative outcome.
enum MosaicCustomerRestoreProviderOutcome {
  completed('completed'),
  noPurchasesFound('no_purchases_found'),
  cancelled('cancelled'),
  failed('failed'),
  unsupported('unsupported'),
  notAttempted('not_attempted');

  const MosaicCustomerRestoreProviderOutcome(this.wireValue);

  final String wireValue;
}

enum MosaicCustomerRestoreStageName {
  providerRestore,
  observationHandoff,
  authoritativeSync,
  completed,
}

/// One observable step of a restore. The example application renders these;
/// support uses them to tell "the store found nothing" apart from "Mosaic has
/// not validated it yet".
final class MosaicCustomerRestoreStage {
  MosaicCustomerRestoreStage({
    required this.name,
    required this.detail,
    required DateTime at,
  }) : at = at.toUtc();

  final MosaicCustomerRestoreStageName name;
  final String detail;
  final DateTime at;
}

/// The outcome of `restorePurchasesAndSync`, on two independent axes.
sealed class MosaicCustomerRestoreResult {
  MosaicCustomerRestoreResult({
    required this.providerOutcome,
    required DateTime requestedAt,
    required Iterable<MosaicCustomerRestoreStage> stages,
  })  : requestedAt = requestedAt.toUtc(),
        stages = List.unmodifiable(stages);

  MosaicCustomerRestoreOutcome get outcome;

  final MosaicCustomerRestoreProviderOutcome providerOutcome;
  final DateTime requestedAt;
  final List<MosaicCustomerRestoreStage> stages;
}

/// The only success. It exists only once an accepted snapshot at a higher
/// version reflects the restore: that snapshot is the evidence that makes the
/// outcome authoritative rather than hopeful.
final class MosaicCustomerEntitlementsRestored
    extends MosaicCustomerRestoreResult {
  MosaicCustomerEntitlementsRestored({
    required super.providerOutcome,
    required super.requestedAt,
    required super.stages,
    required this.snapshotVersion,
    required DateTime completedAt,
    required this.snapshot,
  })  : completedAt = completedAt.toUtc(),
        super();

  @override
  MosaicCustomerRestoreOutcome get outcome =>
      MosaicCustomerRestoreOutcome.restored;

  final int snapshotVersion;
  final DateTime completedAt;
  final MosaicCustomerEntitlementSnapshot snapshot;
}

/// The provider restored purchases and Mosaic has not yet confirmed them. This
/// is the honest answer inside the poll bound, and it is not a failure.
final class MosaicCustomerRestoreValidationPending
    extends MosaicCustomerRestoreResult {
  MosaicCustomerRestoreValidationPending({
    required super.providerOutcome,
    required super.requestedAt,
    required super.stages,
    required this.pendingValidationCount,
    required this.uncertainty,
  }) : super();

  @override
  MosaicCustomerRestoreOutcome get outcome =>
      MosaicCustomerRestoreOutcome.validationPending;

  final int pendingValidationCount;
  final MosaicCustomerUncertainty uncertainty;
}

/// The provider found nothing further to restore and Mosaic agrees.
final class MosaicCustomerRestoreNoAdditionalPurchases
    extends MosaicCustomerRestoreResult {
  MosaicCustomerRestoreNoAdditionalPurchases({
    required super.providerOutcome,
    required super.requestedAt,
    required super.stages,
  }) : super();

  @override
  MosaicCustomerRestoreOutcome get outcome =>
      MosaicCustomerRestoreOutcome.noAdditionalPurchases;
}

/// No Billing Customer could be resolved, so there is nowhere to attach the
/// restore. Access is not claimed in either direction.
final class MosaicCustomerRestoreIdentityUnresolved
    extends MosaicCustomerRestoreResult {
  MosaicCustomerRestoreIdentityUnresolved({
    required super.providerOutcome,
    required super.requestedAt,
    required super.stages,
    required this.uncertainty,
  }) : super();

  @override
  MosaicCustomerRestoreOutcome get outcome =>
      MosaicCustomerRestoreOutcome.identityUnresolved;

  final MosaicCustomerUncertainty uncertainty;
}

final class MosaicCustomerRestoreProviderUnavailable
    extends MosaicCustomerRestoreResult {
  MosaicCustomerRestoreProviderUnavailable({
    required super.providerOutcome,
    required super.requestedAt,
    required super.stages,
    required this.uncertainty,
  }) : super();

  @override
  MosaicCustomerRestoreOutcome get outcome =>
      MosaicCustomerRestoreOutcome.providerUnavailable;

  final MosaicCustomerUncertainty uncertainty;
}

final class MosaicCustomerRestoreFailed extends MosaicCustomerRestoreResult {
  MosaicCustomerRestoreFailed({
    required super.providerOutcome,
    required super.requestedAt,
    required super.stages,
    required this.uncertainty,
    this.reasonCode,
  }) : super();

  @override
  MosaicCustomerRestoreOutcome get outcome =>
      MosaicCustomerRestoreOutcome.failed;

  final MosaicCustomerUncertainty uncertainty;
  final String? reasonCode;
}

typedef MosaicCustomerRestoreClock = DateTime Function();
typedef MosaicCustomerRestoreDelay = Future<void> Function(Duration duration);

DateTime _systemClock() => DateTime.now().toUtc();

Future<void> _wait(Duration duration) => Future<void>.delayed(duration);

/// Composes a native restore, the Transaction Observation handoff, and a
/// bounded authoritative poll into one honest multi-stage result.
///
/// The Flutter Commerce Provider contract exposes no provider transaction
/// references at restore, so references reach Mosaic through the Commerce
/// Provider update stream the SDK already bridges into the observation
/// runtime. This coordinator flushes that queue rather than synthesizing
/// references it does not have.
final class MosaicCustomerRestoreCoordinator {
  MosaicCustomerRestoreCoordinator({
    required this.purchaseProvider,
    required this.entitlements,
    this.observations,
    this.clock = _systemClock,
    this.pollAttempts = mosaicCustomerRestorePollAttempts,
    this.pollBudget = mosaicCustomerRestorePollBudget,
    this.delay = _wait,
  });

  final MosaicPurchaseProvider purchaseProvider;
  final MosaicCustomerEntitlementRuntime entitlements;
  final MosaicTransactionObservationRuntime? observations;
  final MosaicCustomerRestoreClock clock;
  final int pollAttempts;
  final Duration pollBudget;
  final MosaicCustomerRestoreDelay delay;

  Future<MosaicCustomerRestoreResult> restorePurchasesAndSync() async {
    final requestedAt = clock().toUtc();
    final stages = <MosaicCustomerRestoreStage>[];
    final baseline = entitlements.snapshot?.snapshotVersion ?? 0;

    final MosaicRestoreResult providerResult;
    try {
      providerResult = await purchaseProvider.restore();
    } on Object {
      stages.add(_stage(
        MosaicCustomerRestoreStageName.providerRestore,
        'The provider restore threw.',
      ));
      return MosaicCustomerRestoreFailed(
        providerOutcome: MosaicCustomerRestoreProviderOutcome.failed,
        requestedAt: requestedAt,
        stages: stages,
        uncertainty: _uncertainty(
          MosaicCustomerUncertaintyReason.providerUnavailable,
          requestedAt,
        ),
        reasonCode: 'entitlements.restore.providerThrew',
      );
    }
    final providerOutcome = _providerOutcome(providerResult);
    stages.add(_stage(
      MosaicCustomerRestoreStageName.providerRestore,
      providerOutcome.wireValue,
    ));

    switch (providerOutcome) {
      case MosaicCustomerRestoreProviderOutcome.cancelled:
      case MosaicCustomerRestoreProviderOutcome.notAttempted:
      case MosaicCustomerRestoreProviderOutcome.unsupported:
        return MosaicCustomerRestoreFailed(
          providerOutcome: providerOutcome,
          requestedAt: requestedAt,
          stages: stages,
          uncertainty: _uncertainty(
            MosaicCustomerUncertaintyReason.missingFact,
            requestedAt,
          ),
          reasonCode: 'entitlements.restore.${providerOutcome.name}',
        );
      case MosaicCustomerRestoreProviderOutcome.failed:
        return MosaicCustomerRestoreProviderUnavailable(
          providerOutcome: providerOutcome,
          requestedAt: requestedAt,
          stages: stages,
          uncertainty: _uncertainty(
            MosaicCustomerUncertaintyReason.providerUnavailable,
            requestedAt,
          ),
        );
      case MosaicCustomerRestoreProviderOutcome.completed:
      case MosaicCustomerRestoreProviderOutcome.noPurchasesFound:
        break;
    }

    // The handoff is a trigger for server-side validation and never proof. It
    // can fail without changing what the restore reports.
    var handedOff = 0;
    final runtime = observations;
    if (runtime != null) {
      try {
        final flushed = await runtime.flush();
        handedOff = switch (flushed) {
          MosaicTransactionObservationFlushCompleted(:final sent) => sent,
          _ => 0,
        };
      } on Object {
        handedOff = 0;
      }
    }
    stages.add(_stage(
      MosaicCustomerRestoreStageName.observationHandoff,
      '$handedOff observation(s) submitted for validation.',
    ));

    // Bounded poll. Three attempts across roughly six seconds, identical on
    // every platform, so a user does not wait indefinitely for a projection.
    final interval = Duration(
      microseconds: pollAttempts <= 1
          ? 0
          : pollBudget.inMicroseconds ~/ (pollAttempts - 1),
    );
    for (var attempt = 0; attempt < pollAttempts; attempt += 1) {
      if (attempt > 0 && interval > Duration.zero) await delay(interval);
      final refreshed = await entitlements.refresh();
      final snapshot = entitlements.snapshot;
      stages.add(_stage(
        MosaicCustomerRestoreStageName.authoritativeSync,
        'Attempt ${attempt + 1}: ${_describe(refreshed)}',
      ));
      if (snapshot != null && snapshot.snapshotVersion > baseline) {
        // Success is the accepted snapshot, not the provider's optimism.
        stages.add(_stage(
          MosaicCustomerRestoreStageName.completed,
          'Snapshot ${snapshot.snapshotVersion} reflects the restore.',
        ));
        return MosaicCustomerEntitlementsRestored(
          providerOutcome: providerOutcome,
          requestedAt: requestedAt,
          stages: stages,
          snapshotVersion: snapshot.snapshotVersion,
          completedAt: clock().toUtc(),
          snapshot: snapshot,
        );
      }
    }

    if (providerOutcome ==
        MosaicCustomerRestoreProviderOutcome.noPurchasesFound) {
      // The store has nothing further and Mosaic's state did not move. Those
      // two agreeing is a definite answer.
      return MosaicCustomerRestoreNoAdditionalPurchases(
        providerOutcome: providerOutcome,
        requestedAt: requestedAt,
        stages: stages,
      );
    }
    if (entitlements.snapshot == null &&
        entitlements.diagnostics.token.hasToken == false) {
      return MosaicCustomerRestoreIdentityUnresolved(
        providerOutcome: providerOutcome,
        requestedAt: requestedAt,
        stages: stages,
        uncertainty: _uncertainty(
          MosaicCustomerUncertaintyReason.identityUnresolved,
          requestedAt,
        ),
      );
    }
    return MosaicCustomerRestoreValidationPending(
      providerOutcome: providerOutcome,
      requestedAt: requestedAt,
      stages: stages,
      pendingValidationCount: handedOff,
      uncertainty: _uncertainty(
        MosaicCustomerUncertaintyReason.staleValidation,
        requestedAt,
      ),
    );
  }

  MosaicCustomerRestoreStage _stage(
    MosaicCustomerRestoreStageName name,
    String detail,
  ) =>
      MosaicCustomerRestoreStage(name: name, detail: detail, at: clock());

  MosaicCustomerUncertainty _uncertainty(
    MosaicCustomerUncertaintyReason reason,
    DateTime since,
  ) =>
      MosaicCustomerUncertainty(
        reason: reason,
        since: since,
        expectedResolution: MosaicCustomerExpectedResolution.automaticRetry,
      );

  static String _describe(MosaicCustomerEntitlementRefreshResult result) =>
      switch (result) {
        MosaicCustomerEntitlementUpdated(:final snapshot) =>
          'accepted v${snapshot.snapshotVersion}',
        MosaicCustomerEntitlementUnchanged(:final snapshotVersion) =>
          'unchanged at v$snapshotVersion',
        MosaicCustomerEntitlementRejected(:final reasonCode) =>
          'rejected: $reasonCode',
        MosaicCustomerEntitlementUnavailable(:final reasonCode) =>
          'unavailable: $reasonCode',
      };

  static MosaicCustomerRestoreProviderOutcome _providerOutcome(
    MosaicRestoreResult result,
  ) =>
      switch (result) {
        MosaicRestored() => MosaicCustomerRestoreProviderOutcome.completed,
        MosaicNothingToRestore() =>
          MosaicCustomerRestoreProviderOutcome.noPurchasesFound,
        MosaicRestoreCancelled() =>
          MosaicCustomerRestoreProviderOutcome.cancelled,
        MosaicRestoreProviderUnavailable() =>
          MosaicCustomerRestoreProviderOutcome.failed,
        MosaicRestoreConfigurationUnavailable() =>
          MosaicCustomerRestoreProviderOutcome.notAttempted,
        MosaicRestoreFailed() => MosaicCustomerRestoreProviderOutcome.failed,
        MosaicDetailedRestoreResult(:final outcome) => switch (outcome) {
            MosaicCommerceRecoveryOutcome.restored =>
              MosaicCustomerRestoreProviderOutcome.completed,
            MosaicCommerceRecoveryOutcome.nothingToRestore =>
              MosaicCustomerRestoreProviderOutcome.noPurchasesFound,
            MosaicCommerceRecoveryOutcome.cancelled =>
              MosaicCustomerRestoreProviderOutcome.cancelled,
            MosaicCommerceRecoveryOutcome.providerUnavailable ||
            MosaicCommerceRecoveryOutcome.failed =>
              MosaicCustomerRestoreProviderOutcome.failed,
          },
      };
}

File: sdk/android/mosaic/src/main/kotlin/dev/mosaic/sdk/CustomerRestoreSync.kt
package dev.mosaic.sdk

import java.util.concurrent.atomic.AtomicBoolean
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.buffer
import kotlinx.coroutines.launch

/**
 * What the native provider restore itself did, independently of what Mosaic could conclude from it.
 *
 * The two axes are separate because they answer different questions and routinely disagree: a
 * native restore can complete perfectly while Mosaic has not yet validated the recovered purchases,
 * and reporting that as success would promise an authoritative answer that does not exist yet.
 */
enum class MosaicCustomerRestoreProviderOutcome {
    COMPLETED,
    NO_PURCHASES_FOUND,
    CANCELLED,
    FAILED,
    UNSUPPORTED,
    NOT_ATTEMPTED,
}

/** The outcome of a restore composed with an authoritative synchronization. */
sealed interface MosaicCustomerSyncResult {
    val providerOutcome: MosaicCustomerRestoreProviderOutcome

    /**
     * The only success. It is produced when — and only when — Mosaic accepted a snapshot that
     * reflects the recovery, which is what makes the outcome authoritative rather than hopeful.
     */
    data class AuthoritativeEntitlementsUpdated(
        override val providerOutcome: MosaicCustomerRestoreProviderOutcome,
        val snapshot: MosaicCustomerEntitlementSnapshot,
    ) : MosaicCustomerSyncResult

    /**
     * The device recovered purchases and Mosaic has not confirmed them yet.
     *
     * This is not a failure and must not be presented as one: validation is asynchronous, and the
     * snapshot will advance on its own. It is reported honestly so a host does not tell someone
     * their purchase was restored when Mosaic would still answer `unknown`.
     */
    data class NativeRecoveryCompleted(
        override val providerOutcome: MosaicCustomerRestoreProviderOutcome,
        val validationPending: Boolean = true,
    ) : MosaicCustomerSyncResult

    data class NoAdditionalPurchases(
        override val providerOutcome: MosaicCustomerRestoreProviderOutcome,
    ) : MosaicCustomerSyncResult

    data object Cancelled : MosaicCustomerSyncResult {
        override val providerOutcome = MosaicCustomerRestoreProviderOutcome.CANCELLED
    }

    data class ProviderUnavailable(
        override val providerOutcome: MosaicCustomerRestoreProviderOutcome,
        val diagnosticCode: String,
    ) : MosaicCustomerSyncResult

    data class Failed(
        override val providerOutcome: MosaicCustomerRestoreProviderOutcome,
        val diagnosticCode: String,
    ) : MosaicCustomerSyncResult

    /** The store recovered purchases but Mosaic cannot say whose they are. Never "not entitled". */
    data class CustomerUnavailable(
        override val providerOutcome: MosaicCustomerRestoreProviderOutcome,
        val reason: MosaicCustomerEntitlementUnavailableReason,
    ) : MosaicCustomerSyncResult

    data object SignedOut : MosaicCustomerSyncResult {
        override val providerOutcome = MosaicCustomerRestoreProviderOutcome.NOT_ATTEMPTED
    }
}

/** Cross-platform poll bound: three attempts inside roughly six seconds. */
internal const val MOSAIC_CUSTOMER_RESTORE_POLL_ATTEMPTS = 3
internal const val MOSAIC_CUSTOMER_RESTORE_POLL_INTERVAL_MILLIS = 2_000L

/**
 * Restores through the provider, then waits briefly for Mosaic to catch up.
 *
 * The recovery itself is the existing Commerce Provider Contract path — on Google Play, the
 * `queryPurchases` recovery — and it is untouched here. Acknowledgement in particular is not
 * involved: Google's refund window is governed by the adapter, and nothing in this function delays
 * or re-runs it. Observations reach Mosaic through the Transaction Observation runtime that already
 * exists, so this function adds a wait, not a second submission path.
 *
 * The poll is bounded rather than open-ended because validation is genuinely asynchronous. Waiting
 * longer would make a restore button feel broken; not waiting at all would report
 * `validationPending` for every restore that was about to succeed a second later.
 */
internal suspend fun mosaicRestoreAndSyncCustomerEntitlements(
    runtime: MosaicCustomerEntitlementRuntime,
    provider: MosaicPurchaseProvider,
    pollAttempts: Int = MOSAIC_CUSTOMER_RESTORE_POLL_ATTEMPTS,
    pollIntervalMillis: Long = MOSAIC_CUSTOMER_RESTORE_POLL_INTERVAL_MILLIS,
): MosaicCustomerSyncResult {
    val versionBefore = (runtime.customerEntitlements.value as? MosaicCustomerEntitlementSnapshotState.Available)
        ?.snapshot?.snapshotVersion

    val restore = runCatching { provider.restore() }.getOrElse {
        return MosaicCustomerSyncResult.Failed(
            MosaicCustomerRestoreProviderOutcome.FAILED,
            MosaicDiagnosticCode.RESTORE_FAILED.wireName,
        )
    }

    when (val outcome = restore.providerOutcome()) {
        MosaicCustomerRestoreProviderOutcome.CANCELLED -> return MosaicCustomerSyncResult.Cancelled
        MosaicCustomerRestoreProviderOutcome.FAILED -> return MosaicCustomerSyncResult.Failed(
            outcome,
            restore.diagnosticCode(),
        )
        MosaicCustomerRestoreProviderOutcome.UNSUPPORTED,
        MosaicCustomerRestoreProviderOutcome.NOT_ATTEMPTED,
        -> return MosaicCustomerSyncResult.ProviderUnavailable(outcome, restore.diagnosticCode())
        MosaicCustomerRestoreProviderOutcome.NO_PURCHASES_FOUND -> {
            // Still worth one sync: the store found nothing new on this device, but Mosaic may hold
            // a newer projection from another device or from a server-side notification.
            val synced = runtime.refreshCustomerEntitlements()
            return synced.asUpdated(outcome, versionBefore)
                ?: MosaicCustomerSyncResult.NoAdditionalPurchases(outcome)
        }
        MosaicCustomerRestoreProviderOutcome.COMPLETED -> Unit
    }

    repeat(pollAttempts) { attempt ->
        val result = runtime.refreshCustomerEntitlements()
        when (result) {
            is MosaicCustomerEntitlementSyncResult.SignedOut -> return MosaicCustomerSyncResult.SignedOut
            is MosaicCustomerEntitlementSyncResult.Unauthorized -> return MosaicCustomerSyncResult.CustomerUnavailable(
                MosaicCustomerRestoreProviderOutcome.COMPLETED,
                MosaicCustomerEntitlementUnavailableReason.UNAUTHORIZED,
            )
            else -> Unit
        }
        result.asUpdated(MosaicCustomerRestoreProviderOutcome.COMPLETED, versionBefore)?.let { return it }
        if (attempt < pollAttempts - 1) delay(pollIntervalMillis)
    }

    // The device recovered purchases Mosaic has not validated yet. Saying so is the honest answer;
    // claiming a restore would promise an entitlement the SDK has no evidence for.
    return MosaicCustomerSyncResult.NativeRecoveryCompleted(
        MosaicCustomerRestoreProviderOutcome.COMPLETED,
        validationPending = true,
    )
}

/** An accepted snapshot counts only when it actually advanced past the version the restore started at. */
private fun MosaicCustomerEntitlementSyncResult.asUpdated(
    providerOutcome: MosaicCustomerRestoreProviderOutcome,
    versionBefore: Long?,
): MosaicCustomerSyncResult.AuthoritativeEntitlementsUpdated? {
    val snapshot = (this as? MosaicCustomerEntitlementSyncResult.Updated)?.snapshot ?: return null
    if (versionBefore != null && snapshot.snapshotVersion <= versionBefore) return null
    return MosaicCustomerSyncResult.AuthoritativeEntitlementsUpdated(providerOutcome, snapshot)
}

private fun MosaicRestoreResult.providerOutcome(): MosaicCustomerRestoreProviderOutcome = when (this) {
    is MosaicRestoreResult.Detailed -> when (outcome) {
        MosaicCommerceRecoveryOutcome.RESTORED -> MosaicCustomerRestoreProviderOutcome.COMPLETED
        MosaicCommerceRecoveryOutcome.NOTHING_TO_RESTORE -> MosaicCustomerRestoreProviderOutcome.NO_PURCHASES_FOUND
        MosaicCommerceRecoveryOutcome.CANCELLED -> MosaicCustomerRestoreProviderOutcome.CANCELLED
        MosaicCommerceRecoveryOutcome.PROVIDER_UNAVAILABLE -> MosaicCustomerRestoreProviderOutcome.UNSUPPORTED
        MosaicCommerceRecoveryOutcome.FAILED -> MosaicCustomerRestoreProviderOutcome.FAILED
    }
    is MosaicRestoreResult.Restored -> MosaicCustomerRestoreProviderOutcome.COMPLETED
    MosaicRestoreResult.NothingToRestore -> MosaicCustomerRestoreProviderOutcome.NO_PURCHASES_FOUND
    MosaicRestoreResult.Cancelled -> MosaicCustomerRestoreProviderOutcome.CANCELLED
    is MosaicRestoreResult.ProviderUnavailable -> MosaicCustomerRestoreProviderOutcome.UNSUPPORTED
    is MosaicRestoreResult.Failed -> MosaicCustomerRestoreProviderOutcome.FAILED
}

private fun MosaicRestoreResult.diagnosticCode(): String = when (this) {
    is MosaicRestoreResult.ProviderUnavailable -> diagnosticCode
    is MosaicRestoreResult.Failed -> diagnosticCode
    is MosaicRestoreResult.Detailed -> metadata.diagnostics.firstOrNull()?.code
        ?: MosaicDiagnosticCode.RESTORE_FAILED.wireName
    else -> MosaicDiagnosticCode.RESTORE_FAILED.wireName
}

/**
 * Refreshes authoritative entitlements shortly after a purchase completes.
 *
 * Three properties are load-bearing, and they are the same ones the Transaction Observation runtime
 * protects:
 *
 * 1. **It never blocks or delays a purchase.** This is a *subscriber* to the adapter's update flow
 *    with its own bounded buffer, so a hung, slow, or hostile endpoint degrades to a missed refresh
 *    rather than to a stalled purchase. Nothing on the purchase path ever awaits it.
 * 2. **It coalesces.** A purchase commonly produces several updates, and a foreground transition
 *    frequently coincides; one refresh answers all of them.
 * 3. **It changes nothing about entitlement.** The refresh either produces an accepted snapshot or
 *    it does not; a failure here is invisible to the purchase result.
 */
internal class MosaicCustomerPurchaseRefresh(
    private val runtime: MosaicCustomerEntitlementRuntime,
    private val debounceMillis: Long = 1_500,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) {
    private val pending = AtomicBoolean(false)

    fun collect(updates: Flow<MosaicCommerceUpdate>) {
        scope.launch {
            updates
                // An explicit subscriber buffer: the emitting adapter is never suspended by this
                // collector, and the oldest update is dropped rather than the newest, because the
                // newest is the one worth refreshing for.
                .buffer(capacity = 64, onBufferOverflow = BufferOverflow.DROP_OLDEST)
                .collect { update -> observe(update) }
        }
    }

    fun collectPurchaseCompletions(updates: Flow<Unit>) {
        scope.launch {
            updates.buffer(capacity = 64, onBufferOverflow = BufferOverflow.DROP_OLDEST)
                .collect { observePurchaseCompletion() }
        }
    }

    private fun observePurchaseCompletion() {
        if (!pending.compareAndSet(false, true)) return
        scope.launch {
            delay(debounceMillis)
            pending.set(false)
            runCatching { runtime.refreshCustomerEntitlements() }
        }
    }

    private fun observe(update: MosaicCommerceUpdate) {
        if (update.outcome != MosaicCommerceUpdateOutcome.PURCHASED &&
            update.outcome != MosaicCommerceUpdateOutcome.ENTITLEMENTS_CHANGED
        ) {
            return
        }
        observePurchaseCompletion()
    }

    fun close() {
        scope.cancel()
    }
}
